arch/arm/core/cortex_m/mpu/arm_core_mpu.c

```diff
@@ -10,105 +10,294 @@
+ * switch, entering user mode, reconfiguring memory domain, etc.
  *
- * @param thread thread info data structure.
+ * For some MPU architectures, such as the unmodified ARMv8-M MPU,
+ * the function must execute with MPU enabled.
  */
-void configure_mpu_mem_domain(struct k_thread *thread)
+void _arch_configure_dynamic_mpu_regions(struct k_thread *thread)
 {
+	/* Define an array of k_mem_partition objects to hold the configuration
+	 * of the respective dynamic MPU regions to be programmed for
+	 * the given thread. The array of partitions (along with its
+	 * actual size) will be supplied to the underlying MPU driver.
+	 */
+	struct k_mem_partition dynamic_regions[_MAX_DYNAMIC_MPU_REGIONS_NUM];
+
+	u8_t region_num = 0;
+
+#if defined(CONFIG_USERSPACE)
+	/* Memory domain */
 	LOG_DBG("configure thread %p's domain", thread);
-	arm_core_mpu_disable();
-	arm_core_mpu_configure_mem_domain(thread->mem_domain_info.mem_domain);
-	arm_core_mpu_enable();
-}
+	struct k_mem_domain *mem_domain = thread->mem_domain_info.mem_domain;
+	if (mem_domain) {
+		LOG_DBG("configure domain: %p", mem_domain);
+		u32_t num_partitions = mem_domain->num_partitions;
+		struct k_mem_partition partition;
+		int i;
 
-void _arch_mem_domain_configure(struct k_thread *thread)
-{
-	configure_mpu_mem_domain(thread);
+		LOG_DBG("configure domain: %p", mem_domain);
+
+		for (i = 0; i < CONFIG_MAX_DOMAIN_PARTITIONS; i++) {
+			partition = mem_domain->partitions[i];
+			if (partition.size == 0) {
+				/* Zero size indicates a non-existing
+				 * memory partition.
+				 */
+				continue;
+			}
+			LOG_DBG("set region 0x%x 0x%x",
+				partition.start, partition.size);
+			dynamic_regions[region_num].start = partition.start;
+			dynamic_regions[region_num].size = partition.size;
+			dynamic_regions[region_num].attr = partition.attr;
+
+			region_num++;
+			num_partitions--;
+			if (num_partitions == 0) {
+				break;
+			}
+		}
+	}
+	/* Thread user stack */
+	LOG_DBG("configure user thread %p's context", thread);
+	if (thread->arch.priv_stack_start) {
+		u32_t base = (u32_t)thread->stack_obj;
+		u32_t size = thread->stack_info.size;
+#if !defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
+		/* In user-mode the thread stack will include the (optional)
+		 * guard area. For MPUs with arbitrary base address and limit
+		 * it is essential to include this size increase, to avoid
+		 * MPU faults.
+		 */
+		size += thread->stack_info.start - base;
+#endif
+		dynamic_regions[region_num].start = base;
+		dynamic_regions[region_num].size = size;
+		dynamic_regions[region_num].attr = K_MEM_PARTITION_P_RW_U_RW;
+
+		region_num++;
+	}
+#endif /* CONFIG_USERSPACE */
+
+#if defined(CONFIG_MPU_STACK_GUARD)
+	/* Privileged stack guard */
+#if defined(CONFIG_USERSPACE)
+	u32_t guard_start = thread->arch.priv_stack_start ?
+	    (u32_t)thread->arch.priv_stack_start :
+	    (u32_t)thread->stack_obj;
+#else
+	u32_t guard_start = thread->stack_info.start;
+#endif
```

Comment:
Ah, I think I wouldn't do it - especially since arm_core_mpu.c does not contain any other private functions, at all. So readers should "train" to see all implementation in the API functions :)

Reply:
Unless you insist, of course..